src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -2087,6 +2095,69 @@ private BoundExpression MakeNullCheck(SyntaxNode syntax, BoundExpression rewritt
                     type);
         }
 
+        private BoundExpression RewriteBuiltInNativeShiftOperation(
+            BoundBinaryOperator? oldNode,
+            SyntaxNode syntax,
+            BinaryOperatorKind operatorKind,
+            BoundExpression loweredLeft,
+            BoundExpression loweredRight,
+            TypeSymbol type)
+        {
+            // For the predefined operators, the number of bits to shift in `x << count`, `x >> count` or `x >>> count` is computed as follows:
+            // [...]
+            // - When the type of `x` is `nint` or `nuint`, the shift count is given by
+            //   the low-order five bits of `count` on a 32 bit platform, or
+            //   the lower-order six bits of `count` on a 64 bit platform.
+            //   The shift count is computed as `count & (sizeof(nint) * 8 - 1)` or `count & (sizeof(nuint) * 8 - 1)`,
+            //   which is `count & 0x1F` on a 32 bits platform and `count & 0x3F` on a 64 bits platform.
+
+            Debug.Assert(loweredLeft.Type is { });
+            TypeSymbol leftType = loweredLeft.Type;
+            Debug.Assert(leftType.SpecialType is SpecialType.System_IntPtr or SpecialType.System_UIntPtr);
+
+            ConstantValue? rightConstantValue = loweredRight.ConstantValue;
+            Debug.Assert(loweredRight.Type is { });
+            TypeSymbol rightType = loweredRight.Type;
+            Debug.Assert(rightType.SpecialType == SpecialType.System_Int32);
+
+            var oldSyntax = _factory.Syntax;
+            _factory.Syntax = loweredRight.Syntax;
+
+            if (rightConstantValue != null
+                && rightConstantValue.Discriminator == ConstantValueTypeDiscriminator.Int32
+                && (rightConstantValue.Int32Value & 0b100_000) == 0)
+            {
+                // For cases where count doesn't have the sixth bit set, we statically know the masked count.
```

Comment:
>// For cases where count doesn't have the sixth bit set, we statically know the masked count. [](http://example.com/codeflow?start=16&length=93)

It is not clear to me why this is the case. What about shifting by 64 on 128bit platform?

Reply:
`nint` becoming `128-bits` would require revisions to the PE metadata format, the general IL support, and a number of other things. At this point, such a feature would break much existing .NET IL code (particularly given the prevalence of code like `if (IntPtr.Size == 8) { } else { /* assume 32-bit */ }` at some of the lowest layers)

Supporting a 128-bit platform, if it ever happens, is likely many years out. It wouldn't be unreasonable for the current compiler to just pretend it can't happen given the likelihood that it would require a break to the existing metadata format.

It also wouldn't be unreasonable to just guard against "too big shifts" or even emit a warning wave in such scenarios to help tell the user "this might not do what you think it does" (many users think that overshifting results in `0`, like it does for `byte`, `sbyte`, `short`, and `ushort`).

Reply:
I can appreciate the trouble that runtime would have to go through, but I do not see a good reason for the compiler to do so as well (I guess  apart from supporting new metadata format, if that would be even needed). We can easily have correct implementation that doesn't depend on the assumption that 64bit is the biggest size we will ever encounter. I do not see this assumption "helping" compiler today or in the long run.

Reply:
Reverted to simpler logic (only optimize if count is  >= 0 and <= 0x1F). This all isn't worth the effort to optimize for odd cases of negative counts.